### PR TITLE
Check secret is not nil before trying to validate it

### LINKED ISSUE #364: Creating a user with a non-existent secret causes a panic
Create a CassandraDatacenter like this:

```yaml
apiVersion: cassandra.datastax.com/v1beta1
kind: CassandraDatacenter
metadata:
  name: dc1
spec:
  clusterName: cluster1
  serverType: cassandra
  serverVersion: 3.11.7
  managementApiAuth:
    insecure: {}
  size: 3
  storageConfig:
    cassandraDataVolumeClaimSpec:
      storageClassName: server-storage
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 5Gi
  config:
    cassandra-yaml:
      authenticator: org.apache.cassandra.auth.PasswordAuthenticator
      authorizer: org.apache.cassandra.auth.CassandraAuthorizer
      role_manager: org.apache.cassandra.auth.CassandraRoleManager
    jvm-options:
      initial_heap_size: 800M
      max_heap_size: 800M
  users:
  - secretName: does-not-exist
    superuser: true
```

The `validateCassandraUserSecrets` function in `secrets.go` does not check that the secret actually exists. This in turn causes a panic like this:

```
E0121 03:45:47.364093       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 1190 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1570740, 0x25047f0)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1570740, 0x25047f0)
        /usr/local/go/src/runtime/panic.go:969 +0x166
github.com/datastax/cass-operator/operator/pkg/reconciliation.validateCassandraUserSecretContent(0xc0001ac800, 0x0, 0x0, 0x0, 0x1)
        /cass-operator/operator/pkg/reconciliation/secrets.go:231 +0x3a
github.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconciliationContext).validateCassandraUserSecrets(0xc0000c28c0, 0xc0006ebaf0, 0x0, 0x0)
        /cass-operator/operator/pkg/reconciliation/secrets.go:286 +0x1df
github.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconciliationContext).isValid(0xc0000c28c0, 0xc0001ac800, 0xc000319410, 0xc0001b6a10)
        /cass-operator/operator/pkg/reconciliation/handler.go:301 +0x9f
github.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconcileCassandraDatacenter).Reconcile(0xc0007368c0, 0xc000789f99, 0x7, 0xc0007895ec, 0x3, 0x0, 0x0, 0x0, 0x0)
        /cass-operator/operator/pkg/reconciliation/handler.go:245 +0x773
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0001aaa80, 0x15cc3c0, 0xc0006b6900, 0xc000793600)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:256 +0x161
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0001aaa80, 0xc000793700)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:232 +0xae
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0001aaa80)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:211 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00048b2f0)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:152 +0x5f
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00048b2f0, 0x3b9aca00, 0x0, 0xc000453401, 0xc00059db60)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc00048b2f0, 0x3b9aca00, 0xc00059db60)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.5.2/pkg/internal/controller/controller.go:193 +0x305
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x136f4ba]

goroutine 1190 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/pkg/mod/k8s.io/apimachinery@v0.17.4/pkg/util/runtime/runtime.go:55 +0x105
panic(0x1570740, 0x25047f0)
        /usr/local/go/src/runtime/panic.go:969 +0x166
github.com/datastax/cass-operator/operator/pkg/reconciliation.validateCassandraUserSecretContent(0xc0001ac800, 0x0, 0x0, 0x0, 0x1)
        /cass-operator/operator/pkg/reconciliation/secrets.go:231 +0x3a
github.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconciliationContext).validateCassandraUserSecrets(0xc0000c28c0, 0xc0006ebaf0, 0x0, 0x0)
        /cass-operator/operator/pkg/reconciliation/secrets.go:286 +0x1df
github.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconciliationContext).isValid(0xc0000c28c0, 0xc0001ac800, 0xc000319410, 0xc0001b6a10)
        /cass-operator/operator/pkg/reconciliation/handler.go:301 +0x9f
github.com/datastax/cass-operator/operator/pkg/reconciliation.(*ReconcileCassandraDatacenter).Reconcile(0xc0007368c0, 0xc000789f99, 0x7, 0xc0007895ec, 0x3, 0x0, 0x0, 0x0, 0x0)
        /cass-operator/operator/pkg/reconciliation/handler.go:245 +0x773
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0001aaa80, 0x15cc3c0, 0xc0006b6900, 0xc000793600)
```

### PR DESCRIPTION
**Pull request by burmanm**
Before trying to validate the keys inside the Secret, check that the Secret is not nil. Fixes #364

### CODE REVIEW

File: operator/pkg/reconciliation/secrets.go [hunk 1]
```diff
@@ -227,19 +227,23 @@ func (rc *ReconciliationContext) retrieveInternodeCredentialSecretOrCreateDefaul
...
-	}
-	errorPrefix := fmt.Sprintf("Validation failed for user secret: %s", namespacedName.String())
-
-	for _, key := range []string{"username", "password"} {
-		value, ok := secret.Data[key]
-		if !ok {
-			errs = append(errs, fmt.Errorf("%s Missing key: %s", errorPrefix, key))
-		} else if !utf8.Valid(value) {
-			errs = append(errs, fmt.Errorf("%s Key did not have valid utf8 value: %s", errorPrefix, key))
+	if secret != nil {
```
Comment (jimdickinson) [inline operator/pkg/reconciliation/secrets.go:230]: nice find! I'm curious how we got here... did we check if one secret exists and then assume multiple existed?

File: operator/pkg/reconciliation/secrets.go [hunk 2]
```diff
@@ -227,19 +227,23 @@ func (rc *ReconciliationContext) retrieveInternodeCredentialSecretOrCreateDefaul
...
+		for _, key := range []string{"username", "password"} {
+			value, ok := secret.Data[key]
+			if !ok {
+				errs = append(errs, fmt.Errorf("%s Missing key: %s", errorPrefix, key))
+			} else if !utf8.Valid(value) {
+				errs = append(errs, fmt.Errorf("%s Key did not have valid utf8 value: %s", errorPrefix, key))
+			}
 		}
+	} else {
+		errs = append(errs, fmt.Errorf("Validation failed due to a missing secret"))
```
Comment (jimdickinson) [inline operator/pkg/reconciliation/secrets.go:246]: @burmanm am I right that our generic error handling will pull this string onto an event on the CassDC you can see with `kubectl describe`?

**Review by jimdickinson — APPROVED** [review]
lgtm! thank you! just a couple follow up questions

File: operator/pkg/reconciliation/secrets.go (hunk 1 above)
Comment (burmanm) [reply operator/pkg/reconciliation/secrets.go:230]: I don't see any being checked, there was a parameter in the test to check it, but commit 9b376e8be93976a0a344bcda2e417aa90dd9758f removed those tests (and broke the functionality).

File: operator/pkg/reconciliation/secrets.go (hunk 2 above)
Comment (burmanm) [reply operator/pkg/reconciliation/secrets.go:246]: No, currently only the first validation issue is returned as error, rest are ignored:

https://github.com/datastax/cass-operator/blob/master/operator/pkg/reconciliation/handler.go#L306

Comment (burmanm) [reply operator/pkg/reconciliation/secrets.go:246]: But yes, that one error is pushed to the recorder:

https://github.com/datastax/cass-operator/blob/master/operator/pkg/reconciliation/handler.go#L247

**Comment by burmanm** [comment]
@jimdickinson Something that prevents merging?

**Comment by jimdickinson** [comment]
nope, just waiting some minute for Jenkins to be ready to test master with this PR, will merge in 20 mins or so, sorry I didn't stay on this

**Comment by jsanda** [comment]
Thanks @jimdickinson and @burmanm 😄

**Pull request merged**